lean_source/06_Abstract_Algebra/source_02_Algebraic_Structures.lean

```diff
@@ -0,0 +1,676 @@
+of the :numref:`section_structures`.
+An algebraic structure is a specification of a bundle of data satisfying
+some axiomatic hypotheses, and we saw in the :numref:`section_structures` that
+this is exactly what the ``structure`` command is designed to accommodate.
+It's a marriage made in heaven!
+
+Given a data type ``α``, we can define the group structure on ``α``
+as follows.
+EXAMPLES: -/
+-- QUOTE:
+structure group₁ (α : Type*) :=
+(mul: α → α → α)
+(one: α)
+(inv: α → α)
+(mul_assoc : ∀ x y z : α, mul (mul x y) z = mul x (mul y z))
+(mul_one: ∀ x : α, mul x one = x)
+(one_mul: ∀ x : α, mul x one = x)
+(mul_left_inv : ∀ x : α, mul (inv x) x = one)
+-- QUOTE.
+
+-- OMIT: TODO: explain the extends command later, and also redundant inheritance
+/- TEXT:
+Notice that the type ``α`` is a *parameter* in the definition of ``group₁``.
+So you should think of an object ``struc : group₁ α`` as being
+a group structure on ``α``.
+We saw in :numref:`proving_identities_in_algebraic_structures`
+that the counterpart ``mul_right_inv`` to ``mul_left_inv``
+follows from the other group axioms, so there is no need
+to add it to the definition.
+
+This definition of a group is similar to the definition of ``group`` in
+mathlib,
+and we have chosen the name ``group₁`` to distinguish our version.
+If you write ``#check group`` and ctrl-click on the definition,
+you will see that the mathlib version of ``group`` is defined to
+extend another structure; we will explain how to do that later.
+If you type ``#print group`` you will also see that the mathlib
+version of ``group`` has a number of extra fields.
+For reasons we will explain later, sometimes it is useful to add
+redundant information to a structure,
+so that there are additional fields for objects and functions
+that can be defined from the core
+data. Don't worry about that for now.
+Rest assured that our simplified version ``group₁`` is
+morally the same as the definition of a group that mathlib uses.
+
+It is sometimes useful to bundle
+the type together with the structure, and mathlib also
+contains a definition of a ``Group`` structure that is equivalent to
+the following:
+EXAMPLES: -/
+-- QUOTE:
+structure Group₁ :=
+(α : Type*)
+(str : group₁ α)
+-- QUOTE.
+
+/- TEXT:
+The mathlib version is found in ``algebra.category.Group.basic``,
+and you can ``#check`` it if you add this to the imports at the
+beginning of the examples file.
+
+For reasons that will become clearer below, it is more often
+useful to keep the type ``α`` separate from the structure ``group α``.
+We refer to the two objects together as a *partially bundled structure*,
+since the representation combines most, but not all, of the components
+into one structure. It is common in mathlib
+to use capital roman letters like ``G`` for a type
+when it is used as the carrier type for a group.
+
+Let's construct a group, which is to say, an instance of a ``group₁`` structure.
```

Comment:
This is a terminological issue we should talk about. Computer scientists (and the Lean terminology) use the word "structure" for the abstract specification (i.e. the type), and "instance" for particular instantiations (i.e. elements of the type). So they would talk about "the ring structure" and say that the real numbers are "an instance of the ring structure."

Mathematicians don't really have a word for the abstract specification, and they call the instance a "structure". So a mathematician would talk about "the real number structure" and say that the real number structure is a ring.

We can't avoid the CS terminology entirely because we have keywords `structure` and `instance`. Should we explicitly point out the two usages and make sure that readers are clear about them?

Reply:
I'm not sure my point was clear enough here. I mean that we have the Lean keyword `instance`, and the thing you construct at that point in the file does not start with this keyword.